### PR TITLE
Trajectory Services and Savepoint Providers

### PR DESCRIPTION
**Pull request by teubert**
Added the concept of a Trajectory Service that is used by predictor and the future TrajectoryBased LoadEstimator. Trajectory Service is also a SavePoint Provider. Integrated these to be used by Predictor.

### CODE REVIEW

File: inc/TrajectoryService.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+        /**
+         *  Set a waypoint, creates if waypoint doesn't exist already
+         *  @param  tp  Timepoint for waypoint
+         *  @param  wp  Waypoint (point in 3D space
+         */
+        void setWaypoint(time_point tp, const Point3D& wp);
+
+        /**
+         *  delect a waypoint
```
Comment (jason-watkins) [inline inc/TrajectoryService.h]: Only the most delectable waypoints allowed? 🤣

File: src/EventDrivenTrajectoryService.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        bus.subscribe(this, this->source, MessageId::RouteSetWP);
+    }
+
+    EventDrivenTrajectoryService::~EventDrivenTrajectoryService() {
+        lock_guard guard(m);
+        bus.unsubscribe(this);
+    }
+
+    void EventDrivenTrajectoryService::processMessage(const std::shared_ptr<Message>& message) {
+        unique_lock lock(m, std::chrono::milliseconds(10));
```
Comment (jason-watkins) [inline src/EventDrivenTrajectoryService.cpp]: This should be changed to a lock_guard.

File: src/TrajectoryService.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                auto lon = lastWP->getLongitude() +
+                           (waypoint.second.getLongitude() - lastWP->getLongitude()) * ratio;
+                auto alt = lastWP->getAltitude() +
+                           (waypoint.second.getAltitude() - lastWP->getAltitude()) * ratio;
+                return Point3D(lat, lon, alt);
+            }
+            lastWP = &waypoint.second;
+            lastTime = waypoint.first;
+        }
+        throw std::out_of_range("Cannot exterpolate time after last waypoint");
```
Comment (jason-watkins) [inline src/TrajectoryService.cpp]: Pretty sure exterpolate isn't a word

File: inc/EventDrivenTrajectoryService.h [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace PCOE {
+    /**
+     * Maintains knowledge of the route (series of waypoints) and predicted path for a vehicle.
+     * Provides estimates of a vehicles position at a given time, and path-driven save points.
+     *
+     * @author Christopher Teubert
+     * @since 1.2
+     **/
+    class EventDrivenTrajectoryService : public TrajectoryService, public IMessageProcessor {
```
[the remark this answers is not in the document]
Comment (teubert) [reply inc/EventDrivenTrajectoryService.h]: The only reason I can see is to provide access to getSavePts and getPoint methods without duplicating them (or having to call eventDrivenTrajecotryService.getTrajectoryService().getSavePts()). Is there any reason why it should be a data member instead?

**Pull request merged**